add Java Collection delete support with specs & more java_integration specs

colinsurprenant opened the pull request:
this solves #2261

jsvd commented on spec/lib/logstash/java_integration_spec.rb:
```diff
@@ -0,0 +1,82 @@
+      expect(a.size).to eq(0)
+   end
+
+    it "should yeild block when nothing to delete" do
```
yield

ph commented on spec/lib/logstash/java_integration_spec.rb:
```diff
@@ -0,0 +1,82 @@
+    end
+  end
+
+end
```
@colinsurprenant can we use `subject` for `Java::JavaUtil::ArrayList.new` ?

colinsurprenant replied:
you mean replace with
```ruby
subject { Java::JavaUtil::ArrayList.new }
subject += ["foo", "bar", "foo"]
result = subject.delete("baz") {"boo"}
expect(subject.size).to eq(3)
```

?

ph replied:
This is what I propose and display how subject could be useful
```ruby
context "Java::JavaUtil::Collection" do
  context "when deleting elements" do
    subject { Java::JavaUtil::ArrayList.new("foo", "bar") } #not sure if this is supported.
 
    let(:candidate) { "foo" }
 
    it "returns the object" do
      expect(subject.delete(candidate)).to eq(candidate)
    end
 
    it "updates the size of the array" do
      expect_change { subject.delete(candidate) }.to change { subject.size }.by(1)
    end
 
    it "reorder the items" do
      expect_change { subject.delete(candidate) }.to change { subject.first }.from(candidate).to("bar")
    end
  end
```

colinsurprenant replied:
Ok thanks for the examples, TIL expect_change.
I get the idea but in this specific case, I think this is somewhat over engineering a simple test isn't it?

ph replied:
I don't think it is over engineering:

- `expect_change` is actually a better test here than asserting an exact value.
- Running a simple expectation is fast and isolated http://betterspecs.org/#single
- Line of code from your examples and mine are really close similar.
- Using subject and let make test code easier to refactor, move or reuse via shared helpers.

jsvd replied:
Also line one should be describe, not context:

```ruby
describe "Java::JavaUtil::Collection" do
  context "when deleting elements" do
    it "should  ..." do
    end
  end
end
```

ph commented:
@colinsurprenant This fix is working with me, the tests are running.
But I would take some time to tidy up the test suite a bit.

What I mean by that is:

1. Uses subject for the object under evaluation.
2. Uses let statements for the initialization of the array or other useful variables.
3. It is better to have one expect per example, it provide a better separation of errors and context.

ph commented on spec/lib/logstash/java_integration_spec.rb:
```diff
@@ -0,0 +1,84 @@
+        end
+      end.not_to raise_error
+    end
+  end
```
Could we replace this construct with one of the following?

```ruby
expect(subject).to be_a_kind_of(Hash)
expect(subject).to be_a(Hash)

```

ph replied:
I believe `when Klass` is using kind_of?..

colinsurprenant replied:
no. these 2 tests using `is_a?` and using `case/when` are exactly & specifically what need to be be supported and tested. Here rspec dsl will make thing **not as clear** because the java integration code has been specifically written to enable both the `is_a?` and `case/when` functionalities on the related Java classes.

ph replied:
I didn't know you have to write specific code to support the `case` behavior, I have always assumed it was relying on `kind_of` to do the actual type checking. This believe could come from the days I was looking through the rubinius code.

Could you point me to it to the actual logic? I would like to learn how it is done.

colinsurprenant replied:
it's all in `java_integration.rb`, see https://github.com/elastic/logstash/blob/master/lib/logstash/java_integration.rb#L9 and https://github.com/elastic/logstash/blob/master/lib/logstash/java_integration.rb#L28
and comments

ph replied:
Ty I though it was outside of logstash.

colinsurprenant replied:
also, `kind_of` might be a correct test but these java integration patches were specifically made to solve the `case` and `is_a?` use-cases hence the specific test on `case` and `is_a?`.

ph replied:
Or you could test `===` directly?

colinsurprenant replied:
yes, I could but again, we don't use `===` we use `case` so in that respect, changing `===` is an implementation detail and you normally don't want to test an implementation detail, but the end-result behaviour you want.

purbon commented:
+1 on @ph proposals for refactoring the test suite a bit :-)

ph commented:
Other than my minor comments on the test suite this PR is LGTM.
@colinsurprenant since we want to get this merged ASAP we could create a ticket for the test changes if you want.

suyograo commented on lib/logstash/java_integration.rb:
```diff
@@ -39,3 +39,9 @@ def self.===(other)
+
+module java::util::Collection
+  def delete(o)
+    self.removeAll([o]) ? o : block_given? ? yield : nil
```
why the conversion to single element array? can't we just call `self.remove(o)` ?

colinsurprenant replied:
because `self.remove` removes only the first item equal to `o` in `self` and this is not the Ruby   `Array#delete` contract which should deletes all items from self. `self.removeAll` does that but takes a collection of objects as argument.

colinsurprenant commented:
thanks for your suggestions for making better specs :P

ph commented:
LGTM

colinsurprenant commented:
I decided to refactor the `spec/core/event_spec.rb` to test for the initialization of an `Event` from deserialized json for the case where data is from json, deserialized by JrJackson and contains native Java Collection objects. It showed that the `Array |` (union) operator was missing so I also added it in the java integration module and added corresponding specs.

purbon commented on lib/logstash/java_integration.rb:
```diff
@@ -35,7 +19,42 @@ class Hash
+    self.removeAll([o]) ? o : block_given? ? yield : nil
+  end
+
+  # support the Ruby intersection methos on Java Collection
```
methos should be method, right?

colinsurprenant replied:
right :P

jsvd commented:
LGTM

purbon commented:
LGTM, even if the test suites need a bit more love :-) is out of the scope of this PR, I like the usage of a shared example, well done on this @colinsurprenant.

btw, you need rebase :-)

colinsurprenant commented:
great! :shipit:

colinsurprenant commented:
for the record, different union and intersection implementations benchmarks in https://gist.github.com/colinsurprenant/bc3575efde2743f14666
